### ISSUE TITLE
Consider moving to busy service

### ISSUE THREAD

**Issue by ivankravets** [steelbrain/linter#1355]
Firstly, thanks a lot for your work! I have a few questions related to new 2.0 release:

1. Why do we need "yet another busy"? Hundreds of packages use [busy](https://atom.io/packages/busy) service. Would be good to save space in the status panel. What is more, this "green" dot looks very ugly. Grayed?
![screen shot 2017-03-14 at 13 11 44](https://cloud.githubusercontent.com/assets/1760209/23898077/4fe60bf6-08b8-11e7-9776-00e18ad98c1b.png)
2. If your package depends on other Atom packages and can't work without them, don't ask users which they want to install them or not. As result, the basic "linter" will not work at all. Atom provides API to `apm`, see `atom.packages.getApmPath()`. Them, `apm install linter-ui-default --production --compatible`. See [how do that for PlatformIO IDE](https://github.com/platformio/platformio-atom-ide/blob/develop/lib/installer/stages/atom-dependencies.js).
3. Finally, v2.0 totally broke [linter-gcc](https://github.com/hebaishi/linter-gcc), see [report](https://community.platformio.org/t/linter-v2-0-on-platform-ide-2-0-0a8/1615/1).

P.S: [Downgraded "linter"](https://github.com/platformio/platformio-atom-ide/commit/5506ea13a12d5dde434b77c1c45367460adc0529) for PlatformIO IDE to v1.0

**Comment by steelbrain** [comment]
1. `busy` package was written after busy-signal, I even offered the author to merge together on one package instead of two but he disagreed. Regardless, ew cannot use busy because It doesn't show the history of stuff that's already executed. Plus we are using the builtin Atom animations instead of custom ones.

Yes, we could make the dot gray, good idea!

2. The confirmation dialog is completely optional, and was added in case users decide to have a different ui package, we don't want users to be forced to have the default UI package when they want to use something else so we ask them. See [atom-package-deps](https://github.com/steelbrain/package-deps) for more info.

3. Linter-GCC broke because the initial 2.0.0 release didn't have support for legacy indie APIs, which has since been added in https://github.com/steelbrain/linter/pull/1353 and will be released during the next 24 hours. While it would be ideal to have that package on Indie v2 API, Indie v1 is fine too and will be fixed in the upcoming release

**Comment by ivankravets** [comment]
@noseglid what do you think according to 1. question?

**Comment by steelbrain** [comment]
I would be willing to negotiate a merger with the `busy` package if the author is up to it, I require some features from `busy-signal` that `busy` doesn't support, and `busy-signal` has a nicer API too. So at this point in time, I am not considering a move away from busy-signal, I urge other package authors to use `busy-signal` as well :)

**Comment by steelbrain** [comment]
Linter v2.1.0 is out, you can upgrade now if you want. Closing, but feel free to continue discussing

**Comment by noseglid** [comment]
> I require some features from busy-signal that busy doesn't support

What would those be, specifically?

`busy` does indeed have a history with a configurable backlog. It also has duration.

![screen shot 2017-03-14 at 13 36 46](https://cloud.githubusercontent.com/assets/933880/23900918/50f98bd2-08bb-11e7-82fa-0964052254f0.png)

Also, [higher user adoption](https://atom.io/packages/busy). But that might change now that you've included it into linter.

**Comment by fabianfiorotto** [comment]
How do I turn off the three zeros in the status bar? I don't want to see anything there unless I have errors.

**Comment by ivankravets** [comment]
@steelbrain 

1. Thanks, the latest v2.1.0 works in pair with `linter-gcc`
2. Why did you close this issue? We haven't found solution for "multiple busy".

>  I require some features from busy-signal that busy doesn't support

Would be good if you exlain here. 

> busy-signal has a nicer API too

I've just taken a look at your API. Please don't be offended, but this API is useless. "Nice code/API" !== "Solution for a problem/task". We use @noseglid `busy` package because it allows us to run asynchorniously multiple different tasks and report results to `busy`. Task can finish with success or fail with error. This information will be shown in `busy` history. Each asynchornious task has own ID that is very usful to mark when it is finished.

>  Plus we are using the builtin Atom animations

What do you mean? `loading loading-spinner-tiny`? Really, you use the WebKit's animation frames. The same does `busy`.

> I urge other package authors to use busy-signal as well :)

Why we should use software that doesn't resolve a task but has "nice API"?

> So at this point in time, I am not considering a move away from busy-signal

To be honest, I'm really respect you and what you have done with generic `linter` package. I see that you very like "nice API and source code" but forget about UX. Only you, me and @noseglid know about `busy` and what does it mean. The end user doesn't want to hear something about it. He needs a working result and informative status bar without duplicates.

**Comment by Arcanemagus** [comment]
@fabianfiorotto Please follow https://github.com/steelbrain/linter-ui-default/issues/181 for your request.

**Comment by Arcanemagus** [comment]
Reopening this for the discussion about busy signal.

**Comment by steelbrain** [comment]
I ran a quick [GitHub Code Search](https://github.com/search?p=1&q=consumedServices+busy&ref=searchresults&type=Code&utf8=%E2%9C%93) and found the exact number of consumers for both packages for context:

busy-signal:
- atom-force
- linter-ui-default
- intentions

busy:
- atom-build
- java-classpath-registry
- valhalla
- platformio-ide

The results are pretty laughable, I thought busy-signal would be used by more than 3 packages hahaha. @noseglid We could work together on yours or you could join me on mine, still open on the merger, what's your thoughts?

**Comment by noseglid** [comment]
@steelbrain I'd be happy to collaborate and willing to review and discuss any PRs you might have. I'd like to remain the maintainer of `busy` though.

I'd be delighted if linter used `busy` too.

**Comment by steelbrain** [comment]
I've added you as a collaborator on `busy-signal`, It's no longer my packages, it's our package :)

**Comment by noseglid** [comment]
I see I wasn't clear enough.

I don't have the best impression on how you collaborate with other developers (I remember when linter was taken from the Linter organization). I'm not taking side and have nothing to add to that incident, but I try to avoid ending up in those situations. That's why I have declined the invitation to work on `busy-signal`.

I don't really care much about whether `busy` or `busy-signal` end up being the default choice for Atom. It's such a small package with a very limited feature set and provide little challenge.

At the moment I do not have time to migrate any of my packages from `busy` to `busy-signal` therefore I need to keep it. At least for now. If you have spare time, @steelbrain, and it's all the same to you, feel free to migrate to `busy`. While I won't add you as a collaborator, I will definitely handle any PRs and Issues you might have professionally.

**Comment by ivankravets** [comment]
Guys, let's forget about internal problems and how packages were named or were "lived".

@steelbrain I understand your position and why you want to use "all own packages". However, `busy-signal` isn't ideal package. It looks awful and is unfunctional. 

I don't know personally @noseglid, but @PlatformIO made a few PRs to `build` package. What is more, these PRs were @PlatformIO 's specific. Finally, @noseglid merged all of them and provided all help.

### Conclusion

> I've added you as a collaborator

That is not required step. I prefer when an owner is a main maintainer of the package. In this case, we have single "control" under releases and PRs. 

>  You are free to continue using the busy signal or move over to busy-signal.

@steelbrain, I would be thankful if you migrate to `busy` or add some check in the code if `busy` exists, then use it. Or even, if any busy not exists, don't use `busy notifications`.

> Busy-signal history

![screen shot 2017-03-17 at 13 39 30](https://cloud.githubusercontent.com/assets/1760209/24041672/35df72e0-0b17-11e7-90a9-874f06f96d68.png)

Are you sure that this information is useful? It doesn't make any sense. I don't have an influence on it and can't control it. That is really "internal" information for your `linter` plugin. You can put it to `console.debug()`.

----------

If you don't want to migrate to `busy`, please don't force users to install this unuseful "green" dot. Of course, if you "consumed" `busy-signal`, you can still use it to show these multiple `0s`.

Thanks a lot!

**Comment by steelbrain** [comment]
Stepping back a little, if you don't think those notifications are useful and are mostly debug notifications, why not just uninstall `busy-signal` and when the installer offers, say `Never`.

I think it'll work fine 😄

**Comment by ivankravets** [comment]
> why not just uninstall busy-signal and when the installer offers, say Never

Because we automatically pre-install all dependencies without these annoying "Would you like to install A? OK, now restart. After re-start... Would like to install B? ...".

Our community is embedded/hardware specific. They don't know something about "Busy" or even "Linter". Most of them "JUST USE IDE" and it works from a box.

See current dependencies: https://github.com/platformio/platformio-atom-ide/blob/develop/lib/config.js#L105

As you can see, we automatically pre-install all packages and user doesn't see these: " Would you like a cup of coffee? Restart. How about a tee? Restart. How about to uninstall Atom and switch to NotePad? - YES!!! Awful Atom or PIO IDE!!!".

**Comment by steelbrain** [comment]
Oh God, this is such an uncomfortable position that I'm in. On one hand, I want users to have a busy signal by default so I can't pull it out of dependencies, on the other hand I have a package, not even popular, written by someone who is clearly alienated by me, that I should ditch my own package for, just so a linter consumer is happy.

I think a good solution here would be that Platform IDE forks the default linter ui, replaces busy-signal with busy and publishes it for their own userbase. Part of the reason why UIs were made external was so other people can do things that I can't because I need to target the general userbase.

**Comment by ivankravets** [comment]
I have better solution 😊 

Would be thankful if you remove [this line](https://github.com/steelbrain/linter-ui-default/blob/master/package.json#L61) and add new Atom config variable, such as, `linter-ui-default.useBusy` (bool). You can make it enabled by default. 

Before I'll install `linter-ui-default`, I'll set `linter-ui-default.useBusy=false` and then install `linter-ui-default`. Using `require('atom-package-deps').install('busy-signal', true)` you can install extra dependency if it is enabled.

What do you think?

**Comment by steelbrain** [comment]
Sounds like a fair middle ground

**Comment by steelbrain** [comment]
Fix is up in https://github.com/steelbrain/linter-ui-default/pull/211 and will be a part of the next release


tl;dr, before call to `atom-package-deps` in your package where linter is installed, add `atom.config.set('linter-ui-default.useBusySignal', false)` and linter ui won't install it

**Comment by ivankravets** [comment]
Thanks a lot! I've made the same changes on @PlatformIO IDE side https://github.com/platformio/platformio-atom-ide/commit/08cbcba73e456a043f7db4c1fe9c0a65425577aa
